<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>haplink haplotypes · HapLink.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/cli/haplotypes/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../tutorial/1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="../../tutorial/2-examples/">Kicking the tires (Fake sequences)</a></li><li><a class="tocitem" href="../../tutorial/3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="../../tutorial/4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../variants/">haplink variants</a></li><li><a class="tocitem" href="../consensus/">haplink consensus</a></li><li class="is-active"><a class="tocitem" href>haplink haplotypes</a></li><li><a class="tocitem" href="../sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../../api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="../../api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../../api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../../api/variationcall/">VariationCall</a></li><li><a class="tocitem" href="../../api/consensus/">Consensus Tools</a></li><li><a class="tocitem" href="../../api/psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="../../api/haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="../../api/haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../../api/private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">CLI Reference</a></li><li class="is-active"><a href>haplink haplotypes</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>haplink haplotypes</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/cli/haplotypes.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="haplink-haplotypes"><a class="docs-heading-anchor" href="#haplink-haplotypes"><code>haplink haplotypes</code></a><a id="haplink-haplotypes-1"></a><a class="docs-heading-anchor-permalink" href="#haplink-haplotypes" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="HapLink.haplotypes" href="#HapLink.haplotypes"><code>HapLink.haplotypes</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">haplink haplotypes [options] reference variants bam</code></pre><p>Call haplotypes</p><p><strong>Introduction</strong></p><p>Calls haplotypes based on the linkage disequilibrium between subconsensus variant sites on long reads. Variant sites are chosen based on having a &quot;PASS&quot; filter in the <code>variants</code> file, and linkage is calculated based on the reads present in the <code>bam</code> file. Note this means that haplotypes can be called on a different set of sequences than variants were (e.g. variant calling using high accuracy short-read chemistry like Illumina and haplotype calling using low accuracy long-read chemistry like Oxford Nanopore). There are no guarantees that the <code>variants</code> file and <code>bam</code> file match, so use this feature with caution!</p><p><strong>Arguments</strong></p><ul><li><code>reference</code>: path to the reference genome to call haplotypes against in fasta format. Must   not be gzipped, but does not need to be indexed (have a sidecar fai file). HapLink only   supports single-segment reference genomes: if <code>reference</code> includes more than one   sequence, all but the first will be ignored.</li><li><code>variants</code>: path to the variants file that will define variant sites to call haplotypes   from. Must be in VCF (not BCF) v4 format. <a href="../variants/#HapLink.variants"><code>haplink variants</code></a>   generates a compatible file, although output from other tools can also be used.</li><li><code>bam</code>: alignment file to call variants from. Can be in SAM or BAM format, and does not   need to be sorted or indexed, but variant calling speed will increase significantly if   using a sorted and indexed (has a sidebar bai file) BAM file.</li></ul><p><strong>Flags</strong></p><ul><li><code>--simulated-reads</code>: Use maximum likelihood simulation of long reads based on overlapping   short reads</li></ul><p><strong>Options</strong></p><ul><li><code>--outfile=&lt;path&gt;</code>: The file to write haplotype calls to. If left blank, haplotype calls   are written to standard output.</li><li><code>--consensus-frequency=&lt;float&gt;</code>: The minimum frequency at which a variant must appear to   be considered part of the consensus.</li><li><code>--significance=&lt;float&gt;</code>: The alpha value for statistical significance of haplotype calls.</li><li><code>--depth=&lt;int&gt;</code>: Minimum number of times a variant combination must be observed within the   set of reads to be called a haplotype</li><li><code>--frequency=&lt;float&gt;</code>: The minimum proportion of reads that a variant combination must be   observed within compared to all reads covering its position for that haplotype to be   called</li><li><code>--overlap-min=&lt;int&gt;</code>: The minimum number of bases that must overlap for two short reads   to be combined into one simulated read. Can be negative to indicate a minimum distance   between reads. <strong>Only applies when <code>--simulated-reads</code> is set.</strong></li><li><code>--overlap-max=&lt;int&gt;</code>: The maximum number of bases that may overlap for two short reads to   be combined into one simulated read. Can be negative to indicate a cap on how far two   reads must be apart from one another. Must be greater than <code>--overlap-min</code>. <strong>Only   applies when <code>--simulated-reads</code> is set.</strong></li><li><code>--iterations=&lt;int&gt;</code>: The number of simulated reads to create before calling haplotypes.   <strong>Only applies when <code>--simulated-reads</code> is set.</strong></li><li><code>--seed=&lt;int&gt;</code>: The random seed used for picking short reads to create simulated reads   when using maximum likelihood methods. Leaving unset will use the default Julia RNG and   seed. See   <a href="https://docs.julialang.org/en/v1/stdlib/Random/">Julia&#39;s documentation on randomness</a>   for implementation details. <strong>Only applies when <code>--simulated-reads</code> is set.</strong></li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/de8aed1673fc212817b0b99caaf763db3c467c0e/src/HapLink.jl#L222-L281">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../consensus/">« haplink consensus</a><a class="docs-footer-nextpage" href="../sequences/">haplink sequences »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Sunday 4 June 2023 21:55">Sunday 4 June 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
