<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home ¬∑ HapLink.jl</title><meta name="title" content="Home ¬∑ HapLink.jl"/><meta property="og:title" content="Home ¬∑ HapLink.jl"/><meta property="twitter:title" content="Home ¬∑ HapLink.jl"/><meta name="description" content="Documentation for HapLink.jl."/><meta property="og:description" content="Documentation for HapLink.jl."/><meta property="twitter:description" content="Documentation for HapLink.jl."/><meta property="og:url" content="https://ksumngs.github.io/HapLink.jl/"/><meta property="twitter:url" content="https://ksumngs.github.io/HapLink.jl/"/><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img class="docs-light-only" src="assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href>HapLink.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Welcome"><span>Welcome</span></a></li><li><a class="tocitem" href="#Getting-started"><span>Getting started</span></a></li></ul></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="tutorial/1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="tutorial/2-examples/">Kicking the tires (Fake sequences)</a></li><li><a class="tocitem" href="tutorial/3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="tutorial/4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="cli/variants/">haplink variants</a></li><li><a class="tocitem" href="cli/consensus/">haplink consensus</a></li><li><a class="tocitem" href="cli/haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="cli/sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="api/variationcall/">VariationCall</a></li><li><a class="tocitem" href="api/consensus/">Consensus Tools</a></li><li><a class="tocitem" href="api/psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="api/haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="api/haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="api/private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="HapLink"><a class="docs-heading-anchor" href="#HapLink">HapLink</a><a id="HapLink-1"></a><a class="docs-heading-anchor-permalink" href="#HapLink" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.HapLink" href="#HapLink.HapLink"><code>HapLink.HapLink</code></a> ‚Äî <span class="docstring-category">Module</span></header><section><div><pre><code class="language-julia hljs">HapLink</code></pre><p>Viral haplotype calling via linkage disequilibrium</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/c60514550d066bd695942bde25c3cee64a1e590d/src/HapLink.jl#L1-L5">source</a></section></article><h2 id="Welcome"><a class="docs-heading-anchor" href="#Welcome">Welcome</a><a id="Welcome-1"></a><a class="docs-heading-anchor-permalink" href="#Welcome" title="Permalink"></a></h2><p>Howdy! ü§† And welcome to HapLink! üëã HapLink is a command-line suite of tools to enable the exploration of viral quasispecies within a single metagenomic sample. Every piece eventually builds up to our viral haplotype caller, which uses linkage disequilibrium on long sequencing reads (üí° think <a href="https://nanoporetech.com/">Oxford Nanopore</a> or <a href="https://www.pacb.com/">PacBio HiFi</a>) to identify genetic mutations that are conserved within a single virus particle.</p><p>This manual will cover the different ways of using HapLink, starting with a few tutorials before diving into the details of our reference section.</p><h3 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h3><ul><li><a href="api/consensus/#Consensus-Tools">Consensus Tools</a></li><li><a href="api/haplotype/#Haplotype-Methods">Haplotype Methods</a></li><li><a href="api/haplotypecall/#HaplotypeCall"><code>HaplotypeCall</code></a></li><li><a href="api/private/#Private-API-Reference">Private API Reference</a></li><li><a href="api/psuedoread/#PseudoRead"><code>PseudoRead</code></a></li><li><a href="api/variation/#Variation-Methods">Variation Methods</a></li><li><a href="api/variationcall/#VariationCall"><code>VariationCall</code></a></li><li><a href="api/variationinfo/#VariationInfo"><code>VariationInfo</code></a></li><li><a href="api/variationpileup/#VariationPileup"><code>VariationPileup</code></a></li><li><a href="cli/consensus/#haplink-consensus"><code>haplink consensus</code></a></li><li><a href="cli/haplotypes/#haplink-haplotypes"><code>haplink haplotypes</code></a></li><li><a href="cli/sequences/#haplink-sequences"><code>haplink sequences</code></a></li><li><a href="cli/variants/#variants-cli"><code>haplink variants</code></a></li><li><a href="#HapLink">HapLink</a></li><li class="no-marker"><ul><li><a href="#Welcome">Welcome</a></li><li><a href="#Getting-started">Getting started</a></li></ul></li><li><a href="tutorial/1-install/#install-tutorial">In the beginning</a></li><li class="no-marker"><ul><li><a href="tutorial/1-install/#Bioconda">Bioconda</a></li><li><a href="tutorial/1-install/#Comonicon">Comonicon</a></li></ul></li><li><a href="tutorial/2-examples/#cli-tutorial">Kicking the tires</a></li><li class="no-marker"><ul><li><a href="tutorial/2-examples/#Getting-the-goods">Getting the goods</a></li><li><a href="tutorial/2-examples/#Spot-the-difference">Spot the difference</a></li><li><a href="tutorial/2-examples/#The-general-lay-of-the-land">The general lay of the land</a></li><li><a href="tutorial/2-examples/#The-star-attraction">The star attraction</a></li><li><a href="tutorial/2-examples/#Haplotypes-in-the-Matrix">Haplotypes in the Matrix</a></li><li><a href="tutorial/2-examples/#But,-what-does-it-mean?">But, what does it mean?</a></li></ul></li><li><a href="tutorial/3-other/#integration-tutorial">Playing well with others</a></li><li class="no-marker"><ul><li><a href="tutorial/3-other/#Install-the-other-tools">Install the other tools</a></li><li><a href="tutorial/3-other/#Get-sample-data">Get sample data</a></li><li><a href="tutorial/3-other/#Align-sequences">Align sequences</a></li><li><a href="tutorial/3-other/#Call-variants">Call variants</a></li><li><a href="tutorial/3-other/#Call-haplotypes">Call haplotypes</a></li><li><a href="tutorial/3-other/#Alignment-(again,-but-on-a-bigger-scale)">Alignment (again, but on a bigger scale)</a></li><li><a href="tutorial/3-other/#Phylogeny">Phylogeny</a></li></ul></li><li><a href="tutorial/4-repl/#repl-tutorial">For advanced beginners (REPL mode)</a></li><li class="no-marker"><ul><li><a href="tutorial/4-repl/#Installing-HapLink-...-into-a-project">Installing HapLink ... into a project</a></li><li><a href="tutorial/4-repl/#Finding-the-example-files">Finding the example files</a></li><li><a href="tutorial/4-repl/#Pileup-time">Pileup time</a></li><li><a href="tutorial/4-repl/#But-is-it-real?">But is it real?</a></li><li><a href="tutorial/4-repl/#Rule-of-the-majority">Rule of the majority</a></li><li><a href="tutorial/4-repl/#Bringing-in-the-reads">Bringing in the reads</a></li><li><a href="tutorial/4-repl/#Finding-haplotypes...">Finding haplotypes...</a></li><li><a href="tutorial/4-repl/#...and-deciding-if-they&#39;re-real">...and deciding if they&#39;re real</a></li></ul></li></ul><h2 id="Getting-started"><a class="docs-heading-anchor" href="#Getting-started">Getting started</a><a id="Getting-started-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-started" title="Permalink"></a></h2><p>Ready to dive in? ü§ø Here&#39;s a 30,000-foot view</p><pre><code class="language-bash hljs">curl -fsSL https://install.julialang.org | sh -s -- --yes
source ~/.bashrc
julia \
  --startup-file=no \
  --history-file=no \
  --quiet \
  -e &#39;using Pkg; Pkg.activate(;temp=true); Pkg.add(HapLink)&#39;
echo &#39;export PATH=$HOME/.julia/bin:$PATH&#39; &gt;&gt; $HOME/.bashrc
source ~/.bashrc

wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0-rc1/example/reference.fasta
wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0-rc1/example/sample.bam
wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0-rc1/example/sample.bam.bai

haplink variants \
  reference.fasta \
  sample.bam \
  --significance 0.5 \
  --depth 1 \
  --quality 10.0 \
  --position 0.01 \
  --frequency 0.05 \
  | tee sample.vcf

haplink consensus \
  reference.fasta \
  sample.vcf \
  | tee consensus.fasta

haplink haplotypes \
  reference.fasta \
  sample.vcf \
  sample.bam \
  --frequency 0.75 \
  | tee sample.yml

haplink sequences \
  reference.fasta \
  sample.yml \
  | tee sample.fasta</code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="tutorial/1-install/">In the beginning (Installation) ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 22 December 2023 02:00">Friday 22 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
