<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API Reference · HapLink.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/api/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">HapLink.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li class="is-active"><a class="tocitem" href>API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API Reference</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API Reference</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/api.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="API-Reference"><a class="docs-heading-anchor" href="#API-Reference">API Reference</a><a id="API-Reference-1"></a><a class="docs-heading-anchor-permalink" href="#API-Reference" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="HapLink.Haplotype" href="#HapLink.Haplotype"><code>HapLink.Haplotype</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Haplotype(mutations::AbstractVector{Variant})</code></pre><p>Create an object to describe a group of mutations that appear together. A single <a href="#HapLink.Variant"><code>Variant</code></a> can be passed to create a haplotype of a single variant.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/haplotype.jl#L3-L8">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.Variant" href="#HapLink.Variant"><code>HapLink.Variant</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Variant</code></pre><p>Describes a genomic mutation</p><p>The <code>Variant</code> type is based upon the <a href="https://github.com/samtools/hts-specs#variant-calling-data-files">Variant Call Format v4.2 specification</a>, albeit with imperfect compliance.</p><p><strong>Fields</strong></p><ul><li><code>chromosome::String</code>: An identifier from the reference genome or an angle-bracketed   ID String</li><li><code>position::Int</code>: The reference position</li><li><code>identifier::String</code>: Semicolon-separated list of unique identifiers where available. If   there is no identifier available, then &quot;.&quot;&quot; value should be used.</li><li><code>referencebase::NucleotideSeq</code>: Base at (or before, in case of insertion/deletion)   <code>position</code> in the reference genome</li><li><code>alternatebase::NucleotideSeq</code>: Base this mutation describes at <code>position</code>. Note that   each non-reference allele must be represented by a new <code>Variant</code>, unlike the VCF spec</li><li><code>quality::Number</code>: PHRED-scaled quality score for the assertion made by <code>alternatebase</code></li><li><code>filter::Symbol</code>: Filter status, <code>:PASS</code> is this position has passed all filters. Does Not   yet support multiple filters</li><li><code>info::Dict{String,Any}</code>: Additional information. No validation is made concerning the   keys or values.</li></ul><p><strong>Constructors</strong></p><pre><code class="language-julia hljs">Variant(chromosome::String, position::Int, identifier::String, referencebase::NucleotideSeq,
    alternatebase::NucleotideSeq, quality::Number, filter::Symbol, info::Dict{String,Any})

Variant(data::DataFrameRow)</code></pre><p><code>Variant</code>s can be created from the default constructor, or via a row generated by <a href="#HapLink.transformbamcounts-Tuple{AbstractVector{String}}"><code>transformbamcounts</code></a>.</p><p>See also <a href="#HapLink.Haplotype"><code>Haplotype</code></a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variant.jl#L7-L47">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.baseatreferenceposition-Tuple{XAM.BAM.Record, Int64}" href="#HapLink.baseatreferenceposition-Tuple{XAM.BAM.Record, Int64}"><code>HapLink.baseatreferenceposition</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">baseatreferenceposition(record::BAM.Record, pos::Int)</code></pre><p>Get the base at reference position <code>pos</code> present in the sequence of <code>record</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/sequences.jl#L100-L104">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.callvariants-Tuple{DataFrames.AbstractDataFrame, Int64, Int64, Float64, Float64, Float64}" href="#HapLink.callvariants-Tuple{DataFrames.AbstractDataFrame, Int64, Int64, Float64, Float64, Float64}"><code>HapLink.callvariants</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">callvariants(bamcounts::AbstractDataFrame, D_min::Int, Q_min::Int, x_min::Float64,
    f_min::Float64, α::Float64)</code></pre><p>Based on the aligned basecalls and stats in <code>bamcounts</code>, call variants.</p><p><strong>Arguments</strong></p><ul><li><code>bamcounts::AbstractDataFrame</code>: <code>DataFrame</code> containing the output from <code>bam-readcount</code></li><li><code>D_min::Int</code>: minimum variant depth</li><li><code>Q_min::Int</code>: minimum average PHRED-scaled quality at variant position</li><li><code>x_min::Float64</code>: minimum average fractional distance from read end at variant position</li><li><code>f_min::Float64</code>: minimum frequency of variant</li><li><code>α::Float64</code>: significance level of variants by Fisher&#39;s Exact Test</li></ul><p><strong>Returns</strong></p><ul><li><code>Vector{Variant}</code>: <a href="#HapLink.Variant"><code>Variant</code></a>s that passed all the above filters</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variantcalling.jl#L12-L28">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.countbasestats-Tuple{String, String}" href="#HapLink.countbasestats-Tuple{String, String}"><code>HapLink.countbasestats</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">countbasestats(bamfile::String, reffile::String)</code></pre><p>Count and calculate statistics on the basecalls of the alignment in <code>bamfile</code> to the reference genome in <code>reffile</code>. Returns a <code>DataFrame</code> with stats on every base in every alignment position. See <a href="#HapLink.transformbamcounts-Tuple{AbstractVector{String}}"><code>transformbamcounts</code></a> for a complete description of the output DataFrame schema.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/readcounts.jl#L6-L13">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.findsimulatedoccurrences-Tuple{Haplotype, AbstractString}" href="#HapLink.findsimulatedoccurrences-Tuple{Haplotype, AbstractString}"><code>HapLink.findsimulatedoccurrences</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">function findsimulatedoccurrences(args...; kwargs...)</code></pre><p>Find the number of times a particular haplotype is supported by a maximum likelihood simulation of combining aligned reads in a BAM file.</p><p><strong>Arguments</strong></p><ul><li><code>haplotype::Haplotype</code>: The combination of variants to test the aligned reads for evidence   of</li><li><code>bamfile::AbstractString</code>: The path to a BAM file containing aligned reads to be tested   for instances of <code>haplotype</code></li></ul><p><strong>Keywords</strong></p><ul><li><code>iterations::Integer=1000</code>: The number of times to combine reads and test for the presence   of <code>haplotype</code></li></ul><p><code>findsimulatedoccurrences</code> examines each variant position in <code>haplotype</code> and finds a random read containing that position from <code>bamfile</code>. It will then check if the next variant position is contained on the previous read, and if not pick a new random read that contains that variant position. In this way, it assembles a set of reads that conceivably could have come from the same template strand via maximum likelihood.</p><p>From that set of reads, <code>findsimulatedoccurrences</code> returns an <span>$N$</span>-dimensional matrix where <span>$N$</span> is the number of variant positions in <code>haplotypes</code>. The <span>$1$</span> index position in the <span>$n$</span>th dimension represents the number of times the <span>$n$</span>th variant position was found to have the reference base called, while the <span>$2$</span> index position represents the number of times the <span>$n$</span>th variant position was found to have the alternate base called. E.g. <code>first(findsimulatedoccurrences(...))</code> gives the number of times the all-reference base haplotype was found in the simulation, while <code>findsimulatedoccurrences(...)[end]</code> gives the number of times the all-alternate base haplotype was found.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/haplotypecalling.jl#L87-L117">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.linkage-Tuple{AbstractArray{Int64}}" href="#HapLink.linkage-Tuple{AbstractArray{Int64}}"><code>HapLink.linkage</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">linkage(counts::AbstractArray{Int})</code></pre><p>Calculates the linkage disequilibrium and Chi-squared significance level of a combination of haplotypes whose number of occurrences are given by <code>counts</code>.</p><p><code>counts</code> is an <span>$N$</span>-dimensional array where the <span>$N$</span>th dimension represents the <span>$N$</span>th variant call position within a haplotype. <code>findoccurrences</code> produces such an array.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/haplotypecalling.jl#L198-L206">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.matchvariant-Tuple{BioSequences.NucleotideSeq, Variant}" href="#HapLink.matchvariant-Tuple{BioSequences.NucleotideSeq, Variant}"><code>HapLink.matchvariant</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">matchvariant(base::Union{NucleotideSeq,DNA,AbstractVector{DNA}}, var::Variant)</code></pre><p>Checks if <code>base</code> matches the reference or variant expected in <code>var</code>, and returns a symbol indicating which, if any, it matches.</p><p>Returned values can be <code>:reference</code> for a reference match, <code>:alternate</code> for an alternate match, or <code>:other</code> for no match with the given variant.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/sequences.jl#L70-L78">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.mutate-Tuple{FASTX.FASTA.Record, Haplotype}" href="#HapLink.mutate-Tuple{FASTX.FASTA.Record, Haplotype}"><code>HapLink.mutate</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">function mutate(seq::FASTA.Record, haplotype::Haplotype)
function mutate(seq::NucleotideSeq, haplotype::Haplotype)</code></pre><p>Give the reference sequence <code>seq</code> mutations to match the position and basecalls of <code>haplotype</code>. Returns a new sequence, leaving <code>seq</code> unmodified.</p><p>When mutating a <code>FASTA.Record</code>, the new record is given a new unique identifier and description based on the SHA1 hash of the complete genotype.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/sequences.jl#L114-L123">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.myref2seq-Tuple{BioAlignments.Alignment, Int64}" href="#HapLink.myref2seq-Tuple{BioAlignments.Alignment, Int64}"><code>HapLink.myref2seq</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">myref2seq(aln::Alignment, i::Int)</code></pre><p>Replicates the functionality of BioAlignments <code>ref2seq</code>, but can handle hard clips by effectively removing them for the intent of finding the position.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/sequences.jl#L10-L15">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.phrederror-Tuple{Number}" href="#HapLink.phrederror-Tuple{Number}"><code>HapLink.phrederror</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">phrederror(quality::Number)</code></pre><p>Converts a PHRED33-scaled error number into the expected fractional error of basecall</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variantcalling.jl#L3-L7">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.savevcf-Tuple{AbstractVector{Variant}, AbstractString, AbstractString, Int64, Number, Float64, Float64}" href="#HapLink.savevcf-Tuple{AbstractVector{Variant}, AbstractString, AbstractString, Int64, Number, Float64, Float64}"><code>HapLink.savevcf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">savevcf(vars::AbstractVector{Variant}, savepath::String, refpath::String, D::Int,
    Q::Number, x::Float64, α::Float64)</code></pre><p>Save a VCF file populated with <code>vars</code></p><p><strong>Arguments</strong></p><ul><li><code>vars::AbstractVector{Variant}</code>: <code>Vector</code> of <a href="#HapLink.Variant"><code>Variant</code></a>s to write to file</li><li><code>savepath::AbstractString</code>: path of the VCF file to write to. Will be overwritten</li><li><code>refpath::AbstractString</code>: path of the reference genome used to call variants. The   absolute path will be added to the <code>##reference</code> metadata</li><li><code>D::Int</code>: mimimum variant depth used to filter variants. Will be added as <code>##FILTER</code>   metadata</li><li><code>Q::Number</code>: minimum PHRED quality used to filter variants. Will be added as <code>##FILTER</code>   metadata</li><li><code>x::Float64</code>: minimum fractional read position used to filter variants. Will be added as   <code>##FILTER</code> metadata</li><li><code>α::Float64</code>: Fisher&#39;s Exact Test significance level used to filter variants. Will be   added as <code>##FILTER</code> metadata</li></ul><p>Saves the variants in <code>vars</code> to a VCF file at <code>savepath</code>, adding the reference genome <code>refpath</code>, the depth cutoff <code>D</code>, the quality cutoff <code>Q</code>, the position cutoff <code>x</code>, and the significance cutoff <code>α</code> as metadata.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variantcalling.jl#L71-L94">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.serialize_vcf-Tuple{Variant}" href="#HapLink.serialize_vcf-Tuple{Variant}"><code>HapLink.serialize_vcf</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">serialize_vcf(v::Variant)</code></pre><p>Create a VCF line to represent <code>v</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variant.jl#L158-L162">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.serialize_yaml-Tuple{Variant}" href="#HapLink.serialize_yaml-Tuple{Variant}"><code>HapLink.serialize_yaml</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">serialize_yaml(v::Variant)</code></pre><p>Create a valid YAML representation of <code>v</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/variant.jl#L121-L125">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.sumsliced" href="#HapLink.sumsliced"><code>HapLink.sumsliced</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sumsliced(A::AbstractArray, dim::Int, pos::Int=1)</code></pre><p>Sum all elements that are that can be referenced by <code>pos</code> in the <code>dim</code> dimension of <code>A</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; A = reshape(1:8, 2, 2, 2)
2×2×2 reshape(::UnitRange{Int64}, 2, 2, 2) with eltype Int64:
[:, :, 1] =
 1  3
 2  4

[:, :, 2] =
 5  7
 6  8

julia&gt; sumsliced(A, 2)
14

julia&gt; sumsliced(A, 2, 2)
22</code></pre><p>Heavily inspired by Holy, Tim &quot;Multidimensional algorithms and iteration&quot; <a href="https://julialang.org/blog/2016/02/iteration/#filtering_along_a_specified_dimension_exploiting_multiple_indexes">https://julialang.org/blog/2016/02/iteration/#filtering_along_a_specified_dimension_exploiting_multiple_indexes</a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/haplotypecalling.jl#L227-L254">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.transformbamcounts-Tuple{AbstractVector{String}}" href="#HapLink.transformbamcounts-Tuple{AbstractVector{String}}"><code>HapLink.transformbamcounts</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">transformbamcounts(bamcounts::AbstractVector{String})</code></pre><p>Convert the output from <a href="https://github.com/genome/bam-readcount">bam-readcount</a> to a <code>DataFrame</code>.</p><p><strong>Schema</strong></p><table><tr><th style="text-align: right">Column name</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: right"><code>chr</code></td><td style="text-align: left">Chromosome</td></tr><tr><td style="text-align: right"><code>position</code></td><td style="text-align: left">Position</td></tr><tr><td style="text-align: right"><code>reference_base</code></td><td style="text-align: left">Reference base</td></tr><tr><td style="text-align: right"><code>depth</code></td><td style="text-align: left">Total read depth at <code>position</code></td></tr><tr><td style="text-align: right"><code>base</code></td><td style="text-align: left">Alternate base</td></tr><tr><td style="text-align: right"><code>count</code></td><td style="text-align: left">Number of reads containing <code>base</code> at <code>position</code></td></tr><tr><td style="text-align: right"><code>avg_mapping_quality</code></td><td style="text-align: left">Mean mapping quality</td></tr><tr><td style="text-align: right"><code>avg_basequality</code></td><td style="text-align: left">Mean base quality at <code>position</code></td></tr><tr><td style="text-align: right"><code>avg_se_mapping_quality</code></td><td style="text-align: left">Mean single-ended mapping quality</td></tr><tr><td style="text-align: right"><code>num_plus_strand</code></td><td style="text-align: left">Number of reads on the forward strand (N/A)</td></tr><tr><td style="text-align: right"><code>num_minus_strand</code></td><td style="text-align: left">Number of reads on the reverse strand (N/A)</td></tr><tr><td style="text-align: right"><code>avg_pos_as_fraction</code></td><td style="text-align: left">Average position on the read as a fraction, 0 (end) to 1 (center)</td></tr><tr><td style="text-align: right"><code>avg_num_mismatches_as_fraction</code></td><td style="text-align: left">Average number of mismatches on these reads per base</td></tr><tr><td style="text-align: right"><code>avg_sum_mismatch_qualities</code></td><td style="text-align: left">Average sum of the base qualities of mismatches in the reads</td></tr><tr><td style="text-align: right"><code>num_q2_containing_reads</code></td><td style="text-align: left">Number of reads with q2 runs at the 3&#39; end</td></tr><tr><td style="text-align: right"><code>avg_distance_to_q2_start_in_q2_reads</code></td><td style="text-align: left">Average distance of position (as fraction of unclipped read length) to the start of the q2 run</td></tr><tr><td style="text-align: right"><code>avg_clipped_length</code></td><td style="text-align: left">Average clipped read length</td></tr><tr><td style="text-align: right"><code>avg_distance_to_effective_3p_end</code></td><td style="text-align: left">Average distance to the 3&#39; end of the read (as fraction of unclipped read length)</td></tr></table></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/413fab1dcc696dc4ebc41fec1ef90e98aa937df2/src/readcounts.jl#L29-L57">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Home</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.10 on <span class="colophon-date" title="Tuesday 7 December 2021 21:07">Tuesday 7 December 2021</span>. Using Julia version 1.7.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
