<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Plays well with others (External tools) · HapLink.jl</title><meta name="title" content="Plays well with others (External tools) · HapLink.jl"/><meta property="og:title" content="Plays well with others (External tools) · HapLink.jl"/><meta property="twitter:title" content="Plays well with others (External tools) · HapLink.jl"/><meta name="description" content="Documentation for HapLink.jl."/><meta property="og:description" content="Documentation for HapLink.jl."/><meta property="twitter:description" content="Documentation for HapLink.jl."/><meta property="og:url" content="https://ksumngs.github.io/HapLink.jl/tutorial/3-other/"/><meta property="twitter:url" content="https://ksumngs.github.io/HapLink.jl/tutorial/3-other/"/><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/tutorial/3-other/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="../2-examples/">Kicking the tires (Fake sequences)</a></li><li class="is-active"><a class="tocitem" href>Plays well with others (External tools)</a><ul class="internal"><li><a class="tocitem" href="#Install-the-other-tools"><span>Install the other tools</span></a></li><li><a class="tocitem" href="#Get-sample-data"><span>Get sample data</span></a></li><li><a class="tocitem" href="#Align-sequences"><span>Align sequences</span></a></li><li><a class="tocitem" href="#Call-variants"><span>Call variants</span></a></li><li><a class="tocitem" href="#Call-haplotypes"><span>Call haplotypes</span></a></li><li><a class="tocitem" href="#Alignment-(again,-but-on-a-bigger-scale)"><span>Alignment (again, but on a bigger scale)</span></a></li><li><a class="tocitem" href="#Phylogeny"><span>Phylogeny</span></a></li></ul></li><li><a class="tocitem" href="../4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../../cli/variants/">haplink variants</a></li><li><a class="tocitem" href="../../cli/consensus/">haplink consensus</a></li><li><a class="tocitem" href="../../cli/haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="../../cli/sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../../api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="../../api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../../api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../../api/variationcall/">VariationCall</a></li><li><a class="tocitem" href="../../api/consensus/">Consensus Tools</a></li><li><a class="tocitem" href="../../api/psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="../../api/haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="../../api/haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../../api/private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Plays well with others (External tools)</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Plays well with others (External tools)</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/tutorial/3-other.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="integration-tutorial"><a class="docs-heading-anchor" href="#integration-tutorial">Playing well with others: Combining HapLink with external tools</a><a id="integration-tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#integration-tutorial" title="Permalink"></a></h1><p>HapLink is not a one-man show: it knows how to cooperate with other tools! In this tutorial, we&#39;ll let HapLink do the haplotype calling, but use other tools to go from reads to variant calls, and from haplotypes to phylogenies.</p><ul><li><a href="#integration-tutorial">Playing well with others: Combining HapLink with external tools</a></li><li class="no-marker"><ul><li><a href="#Install-the-other-tools">Install the other tools</a></li><li><a href="#Get-sample-data">Get sample data</a></li><li><a href="#Align-sequences">Align sequences</a></li><li><a href="#Call-variants">Call variants</a></li><li><a href="#Call-haplotypes">Call haplotypes</a></li><li><a href="#Alignment-(again,-but-on-a-bigger-scale)">Alignment (again, but on a bigger scale)</a></li><li><a href="#Phylogeny">Phylogeny</a></li></ul></li></ul><h2 id="Install-the-other-tools"><a class="docs-heading-anchor" href="#Install-the-other-tools">Install the other tools</a><a id="Install-the-other-tools-1"></a><a class="docs-heading-anchor-permalink" href="#Install-the-other-tools" title="Permalink"></a></h2><p>Before we get too far, let&#39;s make sure that we actually have all of the tools that we need. Let&#39;s create a brand new conda environment.</p><pre><code class="language-bash hljs">conda create \
  -n haplink-tutorial-3 \
  -c bioconda \
  -c conda-forge \
  wget=1.20 \
  sra-tools=3.0 \
  entrez-direct=16.2 \
  samtools=1.17 \
  minimap2=2.26 \
  lofreq=2.1 \
  haplink \
  mafft=7.520 \
  raxml-ng=1.2 \
  -y
conda activate haplink-tutorial-3</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>None</li></ul></div></div><h2 id="Get-sample-data"><a class="docs-heading-anchor" href="#Get-sample-data">Get sample data</a><a id="Get-sample-data-1"></a><a class="docs-heading-anchor-permalink" href="#Get-sample-data" title="Permalink"></a></h2><p>We&#39;ll pull some of the validation data data from NCBI. First, the reference genome from GenBank.</p><pre><code class="language-bash hljs">esearch \
    -db nucleotide \
    -query &quot;NC_036618.1&quot; \
  | efetch \
    -format fasta \
  &gt; idv4.fasta</code></pre><p>Next, we&#39;ll download one of the pools from the validation set from SRA.</p><pre><code class="language-bash hljs">fasterq-dump --concatenate-reads -X &quot;SRR24796010&quot; | gzip &gt; &quot;IDV-Aug2022-P2.fastq.gz&quot;</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>idv4.fasta</li><li>IDV-Aug2022-P2.fastq.gz</li></ul></div></div><h2 id="Align-sequences"><a class="docs-heading-anchor" href="#Align-sequences">Align sequences</a><a id="Align-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#Align-sequences" title="Permalink"></a></h2><p>We have a set of Nanopore reads and a reference genome to go with them. We&#39;ll use <a href="https://doi.org/10.1093/bioinformatics/bty191">minimap2</a> to align the reads to reference. minimap2 requires the <code>-a</code> flag to output in SAM format, and uses the <code>-x</code> flag to tweak the settings for optimal Nanopore alignment. We then run those reads through <code>samtools sort</code> and <code>samtools index</code> to reduce the computational load needed to find reads by our downstream tools, and <code>samtools view -b</code> to convert the SAM file into a compressed BAM file.</p><pre><code class="language-bash hljs">minimap2 \
    -t $(($(nproc)/2)) \
    -ax map-pb \
    idv4.fasta \
    IDV-Aug2022-P2.fastq.gz \
  | samtools sort \
  | samtools view \
    -@ $(($(nproc)/2)) \
    -b \
    -h \
  &gt; IDV-Aug2022-P2.bam
samtools index IDV-Aug2022-P2.bam</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>IDV-Aug2022-P2.bam</li><li>IDV-Aug2022-P2.bam.bai</li></ul></div></div><h2 id="Call-variants"><a class="docs-heading-anchor" href="#Call-variants">Call variants</a><a id="Call-variants-1"></a><a class="docs-heading-anchor-permalink" href="#Call-variants" title="Permalink"></a></h2><p>People are picky about their variant callers. Even <a href="https://nf-co.re/viralrecon">viralrecon</a> can&#39;t decide which caller to use. If you don&#39;t like HapLink&#39;s built-in variant caller, that&#39;s fine! Here, we&#39;ll use <a href="https://doi.org/10.1093/nar/gks918">LoFreq</a> (another fine variant caller) to decide for us whether nucleotide variations are variants, or just randomness. The only requirement is that the output <em>must</em> be in VCF format (looking at you, <a href="https://doi.org/10.1186/s13059-018-1618-7">iVar</a>) for HapLink to read it.</p><pre><code class="language-bash hljs">lofreq faidx idv4.fasta
lofreq call-parallel \
  --pp-threads $(($(nproc)/2)) \
  --call-indels \
  --ref idv4.fasta \
  IDV-Aug2022-P2.bam \
  --out IDV-Aug2022-P2.vcf</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>IDV-Aug2022-P2.vcf</li></ul></div></div><h2 id="Call-haplotypes"><a class="docs-heading-anchor" href="#Call-haplotypes">Call haplotypes</a><a id="Call-haplotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Call-haplotypes" title="Permalink"></a></h2><p>And we&#39;re back... We now have each piece (reference, alignment, variant calls) needed for HapLink to find haplotypes.</p><pre><code class="language-bash hljs">export JULIA_NUM_THREADS=$(($(nproc)/2))
haplink haplotypes \
  idv4.fasta \
  IDV-Aug2022-P2.vcf \
  IDV-Aug2022-P2.bam \
  &gt; IDV-Aug2022-P2.yml
haplink sequences \
  idv4.fasta \
  IDV-Aug2022-P2.yml \
  &gt; IDV-Aug2022-P2.fasta</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>IDV-Aug2022-P2.yml</li><li>IDV-Aug2022-P2.fasta</li></ul></div></div><h2 id="Alignment-(again,-but-on-a-bigger-scale)"><a class="docs-heading-anchor" href="#Alignment-(again,-but-on-a-bigger-scale)">Alignment (again, but on a bigger scale)</a><a id="Alignment-(again,-but-on-a-bigger-scale)-1"></a><a class="docs-heading-anchor-permalink" href="#Alignment-(again,-but-on-a-bigger-scale)" title="Permalink"></a></h2><p>Now that we have the sequences of every haplotype that HapLink found, we can go farther and see how they compare altogether. We&#39;ll use <a href="https://doi.org/10.1093/nar/gkf436">MAFFT</a> to generate a multiple sequence alignment of every haplotype. By comparing everything to everything, we can start to see patterns that may have been masked by comparing to the reference genome.</p><pre><code class="language-bash hljs">cat idv4.fasta IDV-Aug2022-P2.fasta &gt; sequences.fasta
mafft \
  --thread $(($(nproc)/2)) \
  --auto \
  sequences.fasta \
  &gt; haplotypes.fas</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sequences.fasta</li><li>haplotypes.fas</li></ul></div></div><h2 id="Phylogeny"><a class="docs-heading-anchor" href="#Phylogeny">Phylogeny</a><a id="Phylogeny-1"></a><a class="docs-heading-anchor-permalink" href="#Phylogeny" title="Permalink"></a></h2><p>Although the human brain is <em>reasonably</em> good at picking up pattern differences, we can mathematically determine the similarities between sequences using a phylogeny. We&#39;ll use <a href="https://doi.org/10.1093/bioinformatics/btz305">RAxML-NG</a> to solve the phylogeny for us. Note that RAxML-NG produces <em>a lot</em> of output.</p><pre><code class="language-bash hljs">raxml-ng --all \
  --threads auto{MAX} \
  --msa haplotypes.fas \
  --model GTR+G</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>haplotypes.fas</li><li>haplotypes.fas.raxml.bestModel</li><li>haplotypes.fas.raxml.bestTree</li><li>haplotypes.fas.raxml.bestTreeCollapsed</li><li>haplotypes.fas.raxml.bootstraps</li><li>haplotypes.fas.raxml.log</li><li>haplotypes.fas.raxml.mlTrees</li><li>haplotypes.fas.raxml.rba</li><li>haplotypes.fas.raxml.startTree</li><li>haplotypes.fas.raxml.support</li></ul></div></div><p>You can now import the <code>haplotypes.fas.raxml.bestTree</code> file into a phylogeny viewer to view the tree. A personal favorite around our lab is <a href="https://itol.embl.de/">iTOL</a>, but many others will work.</p><p><img src="../../assets/itol.png" alt="iTOL tree"/></p><ul><li><ul><li><ul><li></li></ul></li></ul></li></ul><p>Now you can do some heavy lifting with HapLink and its friends! You can leave the tutorials and check out the <a href="../../cli/variants/#variants-cli">CLI reference</a>, for an in-depth examination of everything you just learned, or dive into the <a href="../4-repl/#repl-tutorial">next tutorial</a> to learn how to call haplotypes <em>without leaving the Julia REPL.</em></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../2-examples/">« Kicking the tires (Fake sequences)</a><a class="docs-footer-nextpage" href="../4-repl/">For advanced beginners (REPL mode) »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 22 December 2023 19:32">Friday 22 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
