<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Kicking the tires (Fake sequences) · HapLink.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/tutorial/2-examples/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script><link href="../../assets/haplink.css" rel="stylesheet" type="text/css"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../1-install/">In the beginning (Installation)</a></li><li class="is-active"><a class="tocitem" href>Kicking the tires (Fake sequences)</a><ul class="internal"><li><a class="tocitem" href="#Exploring-the-interface"><span>Exploring the interface</span></a></li><li><a class="tocitem" href="#Getting-the-goods"><span>Getting the goods</span></a></li><li><a class="tocitem" href="#Spot-the-difference"><span>Spot the difference</span></a></li><li><a class="tocitem" href="#The-general-lay-of-the-land"><span>The general lay of the land</span></a></li><li><a class="tocitem" href="#The-star-attraction"><span>The star attraction</span></a></li><li><a class="tocitem" href="#If-at-first-you-don&#39;t-succeed"><span>If at first you don&#39;t succeed</span></a></li><li><a class="tocitem" href="#Packaging-it-up"><span>Packaging it up</span></a></li></ul></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../../cli/variants/">haplink variants</a></li></ul></li><li><span class="tocitem">API Reference</span><ul><li><a class="tocitem" href="../../api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="../../api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../../api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../../api/variationcall/">VariationCall</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Kicking the tires (Fake sequences)</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Kicking the tires (Fake sequences)</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/tutorial/2-examples.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="example-sequences"><a class="docs-heading-anchor" href="#example-sequences">Kicking the tires</a><a id="example-sequences-1"></a><a class="docs-heading-anchor-permalink" href="#example-sequences" title="Permalink"></a></h1><p>At this point, we&#39;ll play with the example sequences included <em>gratis</em> with HapLink. No, they don&#39;t represent anything, and they aren&#39;t particularly interesting, but they <strong>do</strong> run fast, so we can get a handle on how the interface and workflow operate.</p><h2 id="Exploring-the-interface"><a class="docs-heading-anchor" href="#Exploring-the-interface">Exploring the interface</a><a id="Exploring-the-interface-1"></a><a class="docs-heading-anchor-permalink" href="#Exploring-the-interface" title="Permalink"></a></h2><p>Go ahead and run <code>haplink</code> in your terminal. You should see a printout similar to that below.</p><pre><code class="language-shellsession hljs">$ haplink
Error: valid sub-commands for command haplink are: consensus, haplotypes, sequences, variants, use -h or --help to check more detailed help info


  haplink v1.0.0-rc1



Usage

  haplink &lt;command&gt;

Commands

  consensus                                                 Convert variant calls to consensus sequence

  haplotypes                                                Call haplotypes

  sequences                                                 Convert haplotype calls into haplotype sequences

  variants                                                  Call variants



Flags

  -h, --help                                                Print this help message.
  --version                                                 Print version.</code></pre><p>HapLink errors out only because you called on it but didn&#39;t give it any work to do. But it did manage to tell us the four tasks that it can do. We&#39;ll do through these commands in <em>logical</em> order. Yeah, it turns out that HapLink sorts the commands in <em>alphabetical</em> order, but if you ran them in that order nothing would happen.</p><h2 id="Getting-the-goods"><a class="docs-heading-anchor" href="#Getting-the-goods">Getting the goods</a><a id="Getting-the-goods-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-the-goods" title="Permalink"></a></h2><p>Let&#39;s get the example files from the code repository. In your terminal, run</p><pre><code class="language-bash hljs">git clone https://github.com/ksumngs/HapLink.jl</code></pre><p>Now let&#39;s switch into the examples directory.</p><pre><code class="language-bash hljs">cd HapLink.jl/example</code></pre><p>What all is available in here?</p><pre><code class="language-shellsession hljs">$ ls -l
.rw-r--r--. ksumngs ksumngs 92 Sat Apr 15 20:47:10 2023 reference.fasta
.rw-r--r--. ksumngs ksumngs 486 Sat Apr 15 20:47:10 2023 sample.bam
.rw-r--r--. ksumngs ksumngs 96 Sat Apr 15 20:47:10 2023 sample.bam.bai</code></pre><p>I see. We have a single sample alignment and its reference genome. It&#39;s a short sequence: we can view the entire thing on the command line.</p><pre><code class="language-shellsession hljs">$ cat reference.fasta
&gt; reference
ACAACTTTATCTCTCTCAACTTCTTCCCTTACTATCCTTCACAACAATCCACACATTACTGCACTTTAAACACTTTTTTA</code></pre><p>If you happen to have <a href="https://www.htslib.org/">samtools</a> installed, you can also view the alignment. Samtools won&#39;t be required for this tutorial, however.</p><pre><code class="language-shellsession hljs">$ # Count the number of lines
$ samtools view sample.bam | wc -l
10

$ samtools view -h sample.bam | head -n6
@HD VN:1.0 SO:coordinate
@SQ SN:reference LN:80
@PG ID:samtools PN:samtools VN:1.13 CL:samtools sort example/sample.sam
@PG ID:samtools.1 PN:samtools PP:samtools VN:1.13 CL:samtools view -h sample.bam
read01 0 reference 6 255 30M * 0 0 TTTATCTGTGTGAACTTCTTGGCTTAGTTT ????????????????????????????5? NM:i:7
read02 0 reference 11 255 30M * 0 0 CTGTGTGAACTTCTTGGCTTAGTATCGTTG ?????????????????????????????? NM:i:8</code></pre><h2 id="Spot-the-difference"><a class="docs-heading-anchor" href="#Spot-the-difference">Spot the difference</a><a id="Spot-the-difference-1"></a><a class="docs-heading-anchor-permalink" href="#Spot-the-difference" title="Permalink"></a></h2><p>In order for HapLink to call haplotypes, it needs to know which sequence differences are due to sequencing errors, and which are due to genetic mutation. This process is known as <em>variant calling</em>, and HapLink comes bundled with a variant caller for just this type of occasion, which requires the reference genome and the alignment BAM file. Since we have both of those, let&#39;s run variant calling now.</p><pre><code class="language-bash hljs">haplink variants reference.fasta sample.bam</code></pre><p>Notice that none of the variants got a <code>PASS</code> filter. In fact, all of them were weeded out by too high of thresholds for depth (remember we only have 10 sequences) and significance. Let&#39;s readjust and save our results.</p><pre><code class="language-bash hljs">haplink \
  variants \
  reference.fasta \
  sample.bam \
  --depth 3 \
  --significance 0.1 \
  | tee sample.vcf</code></pre><p>These settings seemed to work out well. Let&#39;s stick with them and move on.</p><h2 id="The-general-lay-of-the-land"><a class="docs-heading-anchor" href="#The-general-lay-of-the-land">The general lay of the land</a><a id="The-general-lay-of-the-land-1"></a><a class="docs-heading-anchor-permalink" href="#The-general-lay-of-the-land" title="Permalink"></a></h2><p>At this point, we&#39;re going to take a break from haplotype calling and convert those variant calls into a useful summary: the consensus sequence. HapLink, unlike many variant callers, can call the consensus sequence based solely off variant calls. Let&#39;s see that now.</p><pre><code class="language-shellsession hljs">$ haplink consensus reference.fasta sample.vcf
&gt; reference_CONSENSUS
ACAACTTTATCTCTCTCAACTTCTTGGCATAGTTTCGTTGAGAAGAATCGAGAGATTAGTGCAGTTTAAA
CACTTTTTTA</code></pre><h2 id="The-star-attraction"><a class="docs-heading-anchor" href="#The-star-attraction">The star attraction</a><a id="The-star-attraction-1"></a><a class="docs-heading-anchor-permalink" href="#The-star-attraction" title="Permalink"></a></h2><p>And now it&#39;s time for haplotype calling. Before you get your hopes up, there are no <em>true</em> haplotypes in this file. If 10 reads could yield subconsenus mysteries, then bioinformatics would be a super easy job. Alas, we live in the real world, and we&#39;ll have to stretch mathematical constructs to get anything out of these reads.</p><pre><code class="language-bash hljs">haplink \
  haplotypes \
  reference.fasta \
  sample.vcf \
  sample.bam \
  --consensus-frequency 0.75 \
  | tee sample.yml</code></pre><p>You can see that HapLink found only one haplotype in this alignment, but (spoilers!) this isn&#39;t really a haplotype. This is just the consensus sequence, formatted in HapLink&#39;s haplotype scheme.</p><h2 id="If-at-first-you-don&#39;t-succeed"><a class="docs-heading-anchor" href="#If-at-first-you-don&#39;t-succeed">If at first you don&#39;t succeed</a><a id="If-at-first-you-don&#39;t-succeed-1"></a><a class="docs-heading-anchor-permalink" href="#If-at-first-you-don&#39;t-succeed" title="Permalink"></a></h2><p>If you have reads that don&#39;t span the entire genome (like we have here), you can use HapLink&#39;s maximum likelihood simulator to &quot;create&quot; full-length reads by splicing together reads and look for haplotypes on them. Even though we know there aren&#39;t any haplotypes in this sample, let&#39;s get out the simulator and give it a try.</p><pre><code class="language-bash hljs">haplink \
  haplotypes \
  reference.fasta \
  sample.vcf \
  sample.bam \
  --consensus-frequency 0.75 \
  --simulated-reads \
  --iterations 100 \
  --overlap-min 0 \
  --overlap-max 100 \
  | tee sample.ml.yml</code></pre><p>Still nothing, huh? Like I said, no haplotypes here, and simulation can&#39;t change that. Note that simulating full-length reads used <em>a lot</em> more computational power, so you should try to stick with full-length reads when you can!</p><h2 id="Packaging-it-up"><a class="docs-heading-anchor" href="#Packaging-it-up">Packaging it up</a><a id="Packaging-it-up-1"></a><a class="docs-heading-anchor-permalink" href="#Packaging-it-up" title="Permalink"></a></h2><p>HapLink&#39;s haplotype YAML files contain everything needed to recreate the haplotype computation, but they can&#39;t really be used by any other programs. That&#39;s why there&#39;s the <code>sequences</code> command, so haplotype sequences can be saved into FASTA format for use by other tools. Let&#39;s try this now.</p><pre><code class="language-shellsession hljs">$ haplink sequences reference.fasta sample.yml
&gt; CONSENSUS
TTTATCTCTCTCAACTTCTTGGCTTAGTATCGTTGAGAAGAATCGAGAGATTAGTGCAGTTTAAACACTT
TTTTA</code></pre><p>You are now ready to move on to the</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../1-install/">« In the beginning (Installation)</a><a class="docs-footer-nextpage" href="../../cli/variants/">haplink variants »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 24 May 2023 20:37">Wednesday 24 May 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
