<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Kicking the tires (Fake sequences) ¬∑ HapLink.jl</title><meta name="title" content="Kicking the tires (Fake sequences) ¬∑ HapLink.jl"/><meta property="og:title" content="Kicking the tires (Fake sequences) ¬∑ HapLink.jl"/><meta property="twitter:title" content="Kicking the tires (Fake sequences) ¬∑ HapLink.jl"/><meta name="description" content="Documentation for HapLink.jl."/><meta property="og:description" content="Documentation for HapLink.jl."/><meta property="twitter:description" content="Documentation for HapLink.jl."/><meta property="og:url" content="https://ksumngs.github.io/HapLink.jl/tutorial/2-examples/"/><meta property="twitter:url" content="https://ksumngs.github.io/HapLink.jl/tutorial/2-examples/"/><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/tutorial/2-examples/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../1-install/">In the beginning (Installation)</a></li><li class="is-active"><a class="tocitem" href>Kicking the tires (Fake sequences)</a><ul class="internal"><li><a class="tocitem" href="#Getting-the-goods:-extracting-example-files"><span>Getting the goods: extracting example files</span></a></li><li><a class="tocitem" href="#Spot-the-difference:-differentiating-between-sequence-error-and-mutations"><span>Spot the difference: differentiating between sequence error and mutations</span></a></li><li><a class="tocitem" href="#The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls"><span>The general lay of the land: generating consensus sequences from variant calls</span></a></li><li><a class="tocitem" href="#The-star-attraction:-calling-haplotypes-from-sequence-data"><span>The star attraction: calling haplotypes from sequence data</span></a></li><li><a class="tocitem" href="#Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling"><span>Haplotypes in the Matrix: simulating additional reads during haplotype calling</span></a></li><li><a class="tocitem" href="#But,-what-does-it-mean?-Reformatting-into-fasta-format"><span>But, what does it mean? Reformatting into fasta format</span></a></li></ul></li><li><a class="tocitem" href="../3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="../4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../../cli/variants/">haplink variants</a></li><li><a class="tocitem" href="../../cli/consensus/">haplink consensus</a></li><li><a class="tocitem" href="../../cli/haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="../../cli/sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../../api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="../../api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../../api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../../api/variationcall/">VariationCall</a></li><li><a class="tocitem" href="../../api/consensus/">Consensus Tools</a></li><li><a class="tocitem" href="../../api/psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="../../api/haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="../../api/haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../../api/private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Tutorials</a></li><li class="is-active"><a href>Kicking the tires (Fake sequences)</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Kicking the tires (Fake sequences)</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands">ÔÇõ</span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/tutorial/2-examples.md" title="Edit source on GitHub"><span class="docs-icon fa-solid">ÔÅÑ</span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="cli-tutorial"><a class="docs-heading-anchor" href="#cli-tutorial">Starting at the command line</a><a id="cli-tutorial-1"></a><a class="docs-heading-anchor-permalink" href="#cli-tutorial" title="Permalink"></a></h1><p>At this point, we&#39;ll play with the example sequences included <em>gratis</em> üí∞ with HapLink. No, they don&#39;t represent anything, and they aren&#39;t particularly interesting ü•±, but they <strong>do</strong> run fast üèá, so we can get a handle on how the interface and workflow operate.</p><ul><li><a href="#cli-tutorial">Starting at the command line</a></li><li class="no-marker"><ul><li><a href="#Getting-the-goods:-extracting-example-files">Getting the goods: extracting example files</a></li><li><a href="#Spot-the-difference:-differentiating-between-sequence-error-and-mutations">Spot the difference: differentiating between sequence error and mutations</a></li><li><a href="#The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls">The general lay of the land: generating consensus sequences from variant calls</a></li><li><a href="#The-star-attraction:-calling-haplotypes-from-sequence-data">The star attraction: calling haplotypes from sequence data</a></li><li><a href="#Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling">Haplotypes in the Matrix: simulating additional reads during haplotype calling</a></li><li><a href="#But,-what-does-it-mean?-Reformatting-into-fasta-format">But, what does it mean? Reformatting into fasta format</a></li></ul></li></ul><h2 id="Getting-the-goods:-extracting-example-files"><a class="docs-heading-anchor" href="#Getting-the-goods:-extracting-example-files">Getting the goods: extracting example files</a><a id="Getting-the-goods:-extracting-example-files-1"></a><a class="docs-heading-anchor-permalink" href="#Getting-the-goods:-extracting-example-files" title="Permalink"></a></h2><p>Let&#39;s get the example files from the code repository. In your terminal, run</p><pre><code class="language-bash hljs">wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0/example/reference.fasta
wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0/example/sample.bam
wget https://github.com/ksumngs/HapLink.jl/raw/v1.0.0/example/sample.bam.bai</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>reference.fasta</li><li>sample.bam</li><li>sample.bam.bai</li></ul></div></div><h2 id="Spot-the-difference:-differentiating-between-sequence-error-and-mutations"><a class="docs-heading-anchor" href="#Spot-the-difference:-differentiating-between-sequence-error-and-mutations">Spot the difference: differentiating between sequence error and mutations</a><a id="Spot-the-difference:-differentiating-between-sequence-error-and-mutations-1"></a><a class="docs-heading-anchor-permalink" href="#Spot-the-difference:-differentiating-between-sequence-error-and-mutations" title="Permalink"></a></h2><p>In order for HapLink to call haplotypes, it needs to know which sequence differences are due to sequencing errors, and which are due to genetic mutation. This process is known as <em>variant calling</em>, and HapLink comes bundled with a variant caller for just this type of occasion, which requires the reference genome and the alignment BAM file. Since we have both of those, let&#39;s run variant calling now.</p><pre><code class="language-bash hljs">haplink variants reference.fasta sample.bam</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><p><em>None</em></p></div></div><p>HapLink by default outputs to standard output, so the variant calls were printed on your screen instead of saved. That&#39;s okay, though. It&#39;s often good to visually check your variant calls, and in this case, we absolutely needed to. Notice that none of the variants got a <code>PASS</code> filter. In fact, all of them were weeded out by the depth threshold (remember we only have 10 sequences) and significance. Let&#39;s readjust (and save our results this time).</p><pre><code class="language-bash hljs">haplink \
  variants \
  reference.fasta \
  sample.bam \
  --depth 3 \
  --significance 0.1 \
  | tee sample.vcf</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sample.vcf</li></ul></div></div><p>These settings seemed to work out well. Let&#39;s stick with them and move on.</p><h2 id="The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls"><a class="docs-heading-anchor" href="#The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls">The general lay of the land: generating consensus sequences from variant calls</a><a id="The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls-1"></a><a class="docs-heading-anchor-permalink" href="#The-general-lay-of-the-land:-generating-consensus-sequences-from-variant-calls" title="Permalink"></a></h2><p>At this point, we&#39;re going to take a break from haplotype calling and convert those variant calls into a useful summary: the consensus sequence. HapLink can call the consensus sequence based solely off variant calls. Let&#39;s see that now.</p><pre><code class="language-bash hljs">haplink consensus reference.fasta sample.vcf | tee sample.consensus.fasta</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sample.consensus.fasta</li></ul></div></div><h2 id="The-star-attraction:-calling-haplotypes-from-sequence-data"><a class="docs-heading-anchor" href="#The-star-attraction:-calling-haplotypes-from-sequence-data">The star attraction: calling haplotypes from sequence data</a><a id="The-star-attraction:-calling-haplotypes-from-sequence-data-1"></a><a class="docs-heading-anchor-permalink" href="#The-star-attraction:-calling-haplotypes-from-sequence-data" title="Permalink"></a></h2><p>And now it&#39;s time for haplotype calling. Before you get your hopes up, there are no <em>true</em> haplotypes in this file. If 10 reads could manifest subconsenus mysteries, then bioinformatics would be a super easy job. Alas, we live in the real world, and we&#39;ll have to stretch mathematical constructs to get anything out of these reads.</p><pre><code class="language-bash hljs">haplink \
  haplotypes \
  reference.fasta \
  sample.vcf \
  sample.bam \
  --consensus-frequency 0.75 \
  | tee sample.yml</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sample.yml</li></ul></div></div><p>You can see that HapLink found only one haplotype in this alignment, but (spoilers!) this isn&#39;t really a haplotype. This is just the consensus sequence, formatted in HapLink&#39;s haplotype scheme. The first haplotype in any output file is always the consensus sequence.</p><h2 id="Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling"><a class="docs-heading-anchor" href="#Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling">Haplotypes in the Matrix: simulating additional reads during haplotype calling</a><a id="Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling-1"></a><a class="docs-heading-anchor-permalink" href="#Haplotypes-in-the-Matrix:-simulating-additional-reads-during-haplotype-calling" title="Permalink"></a></h2><p>If you have reads that don&#39;t span the entire genome (like we have here), you can use HapLink&#39;s maximum likelihood simulator to &quot;create&quot; full-length reads by splicing together reads and look for haplotypes on them. Even though we know there aren&#39;t any haplotypes in this sample, let&#39;s get out the simulator and give it a try.</p><pre><code class="language-bash hljs">haplink \
  haplotypes \
  reference.fasta \
  sample.vcf \
  sample.bam \
  --consensus-frequency 0.75 \
  --simulated-reads \
  --iterations 100 \
  --overlap-min 0 \
  --overlap-max 100 \
  | tee sample.ml.yml</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sample.ml.yml</li></ul></div></div><p>Still nothing, huh? Like I said, no haplotypes here, and simulation can&#39;t change that. Note that simulating full-length reads used <em>a lot</em> more computational power, so you should try to stick with full-length reads when you can!</p><h2 id="But,-what-does-it-mean?-Reformatting-into-fasta-format"><a class="docs-heading-anchor" href="#But,-what-does-it-mean?-Reformatting-into-fasta-format">But, what does it mean? Reformatting into fasta format</a><a id="But,-what-does-it-mean?-Reformatting-into-fasta-format-1"></a><a class="docs-heading-anchor-permalink" href="#But,-what-does-it-mean?-Reformatting-into-fasta-format" title="Permalink"></a></h2><p>HapLink&#39;s haplotype YAML files contain everything needed to recreate the haplotype computation, but they can&#39;t really be used by any other programs. That&#39;s why there&#39;s the <code>sequences</code> command, so haplotype sequences can be saved into FASTA format for use by other tools. Let&#39;s try this now.</p><pre><code class="language-bash hljs">haplink sequences reference.fasta sample.yml &gt; sample.fasta</code></pre><div class="admonition is-info"><header class="admonition-header">Output</header><div class="admonition-body"><ul><li>sample.fasta</li></ul></div></div><p>The output only contains the consensus sequence. Since we knew that there were no haplotypes in this sample we could have used <code>haplink consensus</code>, instead to get the same result.</p><ul><li><ul><li><ul><li></li></ul></li></ul></li></ul><p>You are now ready to move on to the <a href="../3-other/#integration-tutorial">next tutorial</a>!</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../1-install/">¬´ In the beginning (Installation)</a><a class="docs-footer-nextpage" href="../3-other/">Plays well with others (External tools) ¬ª</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 22 December 2023 19:05">Friday 22 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
