<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>HaplotypeCall · HapLink.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/api/haplotypecall/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../tutorial/1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="../../tutorial/2-examples/">Kicking the tires (Fake sequences)</a></li><li><a class="tocitem" href="../../tutorial/3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="../../tutorial/4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../../cli/variants/">haplink variants</a></li><li><a class="tocitem" href="../../cli/consensus/">haplink consensus</a></li><li><a class="tocitem" href="../../cli/haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="../../cli/sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../variation/">Variation Extensions</a></li><li><a class="tocitem" href="../variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../variationcall/">VariationCall</a></li><li><a class="tocitem" href="../consensus/">Consensus Tools</a></li><li><a class="tocitem" href="../psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="../haplotype/">Haplotype Extensions</a></li><li class="is-active"><a class="tocitem" href>HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Public API Reference</a></li><li class="is-active"><a href>HaplotypeCall</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>HaplotypeCall</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/api/haplotypecall.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="HaplotypeCall"><a class="docs-heading-anchor" href="#HaplotypeCall"><code>HaplotypeCall</code></a><a id="HaplotypeCall-1"></a><a class="docs-heading-anchor-permalink" href="#HaplotypeCall" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="HapLink.HaplotypeCall" href="#HapLink.HaplotypeCall"><code>HapLink.HaplotypeCall</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">HaplotypeCall{S,T}</code></pre><p>A <code>Haplotype</code> that has had linkage disequilibrium calculations performed along with metadata to verify or refute its validity. Very similar to a <a href="../variationcall/#HapLink.VariationCall"><code>VariationCall</code></a>, but for <code>Haplotype</code>s.</p><p><strong>Fields</strong></p><ul><li><code>depth::UInt64</code>: How many sequencing reads contain all the <code>Variation</code>s in <code>haplotype</code></li><li><code>linkage::Float64</code>: The unweighted linkage disequilibrium coefficient of this call</li><li><code>frequency::Float64</code>: How often this haplotype occurs compared to the entire read pool</li><li><code>significance::Float64</code>: The <span>$χ^2$</span> statistical significance (<span>$p$</span>-value) of this call</li><li><code>haplotype::Haplotype{S,T}</code>: The set of <code>Variation</code>s that are inherited together in this   call</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L1-L16">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.HaplotypeCall-Union{Tuple{T}, Tuple{S}, Tuple{SequenceVariation.Haplotype{S, T}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}" href="#HapLink.HaplotypeCall-Union{Tuple{T}, Tuple{S}, Tuple{SequenceVariation.Haplotype{S, T}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}"><code>HapLink.HaplotypeCall</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">HaplotypeCall(
    haplotype::Haplotype{S,T}, reads::AbstractArray{Haplotype{S,T}}
) where {S&lt;:BioSequence,T&lt;:BioSymbol}
HaplotypeCall(
    haplotype::AbstractArray{Variation{S,T}}, reads::AbstractArray{Haplotype{S,T}}
) where {S&lt;:BioSequence,T&lt;:BioSymbol}</code></pre><p>Calls haplotypes by calculating the linkage disequilibrium of <code>haplotype</code> within <code>reads</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L64-L73">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.depth-Tuple{HaplotypeCall}" href="#HapLink.depth-Tuple{HaplotypeCall}"><code>HapLink.depth</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">depth(hc::HaplotypeCall)</code></pre><p>Gets the number of times <code>hc</code> was found</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L25-L29">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.frequency-Tuple{HaplotypeCall}" href="#HapLink.frequency-Tuple{HaplotypeCall}"><code>HapLink.frequency</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">frequency(hc::HaplotypeCall)</code></pre><p>Gets the relative number of times <code>hc</code> was found</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L39-L43">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.haplotype-Tuple{HaplotypeCall}" href="#HapLink.haplotype-Tuple{HaplotypeCall}"><code>HapLink.haplotype</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">haplotype(hc::HaplotypeCall)</code></pre><p>Gets the underlying <code>Haplotype</code> of <code>hc</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L53-L57">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.ishaplotype-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{SequenceVariation.Variation{S, T}}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}" href="#HapLink.ishaplotype-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{SequenceVariation.Variation{S, T}}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}"><code>HapLink.ishaplotype</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">function ishaplotype(
    haplotype::Union{AbstractArray{Variation{S,T}}, Haplotype{S,T}},
    reads::AbstractArray{Haplotype{S,T}};
    frequency::Union{Float64,Nothing}=nothing,
    significance::Union{Float64,Nothing}=nothing,
    depth::Union{Int,Nothing}=nothing,
) where {S&lt;:BioSequence,T&lt;:BioSymbol}</code></pre><p>Determines if a call of <code>haplotype</code> is supported by the sequences in <code>reads</code> based upon the provided keyword criteria.</p><p><strong>Arguments</strong></p><ul><li><code>haplotype::Union{AbstractArray{Variation}, Haplotype}</code>: A <code>Vector</code> of <code>Variation</code>s or a   <code>Haplotype</code> to search for as a haplotype</li><li><code>reads::AbstractArray{Haplotype}</code>: The reads to search for <code>haplotype</code> in</li></ul><p><strong>Keywords</strong></p><ul><li><code>frequency::Union{Float64,Nothing}=nothing</code>: The minimum number of times the entire   <code>haplotype</code> must appear within <code>reads</code> compared to the number of reads to return <code>true</code></li><li><code>significance::Union{Float64,Nothing}=nothing</code>: The <span>$χ^2$</span> significance level (<span>$α$</span>)   of linkage disequilibrium that a haplotype must surpass to return <code>true</code></li><li><code>depth::Union{Int,Nothing}=nothing</code>: The minimum number of times the entire <code>haplotype</code>   must appear within <code>reads</code> to return <code>true</code></li></ul><p><strong>Extended help</strong></p><p>Linkage disequilibrium (<span>$Δ$</span>) is calculated by</p><p class="math-container">\[Δ = P_{reference} - \prod_i P_{ref,i}\]</p><p>where</p><ul><li><span>$P_{reference}$</span> is the probability of finding a read that contains only reference   (consensus) bases</li><li><span>$P_{ref,i}$</span> is the probability of finding a read that contains the reference (consensus)   base for variant <span>$i$</span> within a haplotype</li></ul><p>and the <span>$χ^2$</span> statistic is calculated by</p><p class="math-container">\[χ^2 = \frac{
            Δ^2 n
        }
        {
            \left(\prod_i^N {P_{ref,i} \left(1 - P_{ref,i}\right)}\right)^\frac{2}{N}
        }\]</p><p>where</p><ul><li><span>$N$</span> is the number of <code>Variation</code>s in <code>haplotype</code> (i.e., <code>length(haplotype)</code>)</li><li><span>$n$</span> is the number of total reads sampled (i.e. <code>length(reads)</code>)</li></ul><p>The significance is then calculated from the cumulative <span>$χ^2$</span> distribution function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L96-L150">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.linkage-Tuple{AbstractArray{&lt;:Integer}}" href="#HapLink.linkage-Tuple{AbstractArray{&lt;:Integer}}"><code>HapLink.linkage</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">linkage(counts::AbstractArray{&lt;:Integer})</code></pre><p>Calculates the linkage disequilibrium of a haplotype given its <span>$N$</span>-dimensional contingency matrix, <code>counts</code>.</p><p><code>counts</code> is an <span>$N$</span>-dimensional array where the <span>$N$</span>th dimension represents the <span>$N$</span>th variant call locus within a haplotype. <code>findoccurrences</code> produces such an array.</p><p><strong>Extended help</strong></p><p><code>linkage(::AbstractArray{&lt;:Integer})</code> calculates an unweighted linkage disequilibrium as given by Equation (6) of <a href="https://doi.org/10.1093/genetics/72.1.157">Slatkin (1972)</a>.</p><p class="math-container">\[D(1..N) = E\left( \prod_k^N i_k - P_k \right)\]</p><p>where</p><ul><li><span>$N$</span> is the number of variant loci</li><li><span>$D(1..N)$</span> is the linkage disequilibrium of all <span>$N$</span> variant loci</li><li><span>$E$</span> is an operator returning the arithmetic mean of its argument over every read</li><li><span>$i_k$</span> is a function that returns <span>$1$</span> if the <span>$k$</span>-th locus of the given read contains the reference allele, and returns <span>$0$</span> otherwise.</li><li><span>$P_k$</span> is the probability of any read containing the reference allele in the <span>$k$</span>-th locus, i.e. the frequency at which the reference allele is found within the entire read set at the <span>$k$</span>-th locus</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L282-L310">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.linkage-Tuple{HaplotypeCall}" href="#HapLink.linkage-Tuple{HaplotypeCall}"><code>HapLink.linkage</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">linkage(hc::HaplotypeCall)</code></pre><p>Gets the unweighted linkage disequilibrium coefficient of <code>hc</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L32-L36">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.occurrence_matrix-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{SequenceVariation.Variation{S, T}}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}" href="#HapLink.occurrence_matrix-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{SequenceVariation.Variation{S, T}}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}"><code>HapLink.occurrence_matrix</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">occurrence_matrix(
    haplotype::AbstractArray{Variation{S,T}},
    reads::AbstractArray{Haplotype{S,T}},
) where {S&lt;:BioSequence,T&lt;:BioSymbol}
occurrence_matrix(
    haplotype::Haplotype{S,T},
    reads::AbstractArray{S,T}
) where {S&lt;:BioSequence,T&lt;:BioSymbol}</code></pre><p>Determine how many times the variants in <code>haplotype</code> appear in <code>reads</code> as an <span>$N$</span>- dimensional matrix.</p><p><strong>Arguments</strong></p><ul><li><code>haplotype::AbstractArray{Variation}</code>: A <code>Vector</code> of <code>Variation</code>s or a <code>Haplotype</code> to search   for as a haplotype</li><li><code>reads::AbstractArray{Haplotype}</code>: The reads to search for <code>haplotype</code> in</li></ul><p><strong>Returns</strong></p><ul><li><code>2x2x... Array{Int, length(haplotype)}</code>: An <span>$N$</span>-dimensional matrix where <span>$N$</span> is   the number of variant positions in <code>readmatches</code>. The <span>$1$</span> index position in the   <span>$n$</span>th dimension represents the number of times the <span>$n$</span>th variant position was found   to have the reference base called, while the <span>$2$</span> index position represents the number   of times the <span>$n$</span>th variant position was found to have the alternate base called. E.g.   <code>first(occurrence_matrix(reads))</code> gives the number of times the all-reference base   haplotype was found in <code>reads</code>, while <code>occurrence_matrix(reads)[end]</code> gives the number   of times the all-alternate base haplotype was found.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L225-L252">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.significance-Tuple{AbstractArray{&lt;:Integer}}" href="#HapLink.significance-Tuple{AbstractArray{&lt;:Integer}}"><code>HapLink.significance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">significance(counts::AbstractArray{&lt;:Integer})</code></pre><p>Calculates the <span>$χ^2$</span> significance of a haplotype given its <span>$N$</span>-dimensional contingency matrix, <code>counts</code></p><p>See the documentation on <a href="#HapLink.linkage-Tuple{AbstractArray{&lt;:Integer}}"><code>linkage(::AbstractArray{&lt;:Integer})</code></a> or <a href="#HapLink.occurrence_matrix-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{SequenceVariation.Variation{S, T}}, AbstractArray{SequenceVariation.Haplotype{S, T}}}} where {S&lt;:BioSequences.BioSequence, T&lt;:BioSymbols.BioSymbol}"><code>occurrence_matrix</code></a> for details on how <code>counts</code> should be constructed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L346-L354">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.significance-Tuple{HaplotypeCall}" href="#HapLink.significance-Tuple{HaplotypeCall}"><code>HapLink.significance</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">significance(hc::HaplotypeCall)</code></pre><p>Gets the <span>$χ^2$</span> statistical significance (<span>$p$</span>-value) of <code>hc</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L46-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="HapLink.sumsliced" href="#HapLink.sumsliced"><code>HapLink.sumsliced</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">sumsliced(A::AbstractArray, dim::Int, pos::Int=1)</code></pre><p>Sum all elements that are that can be referenced by <code>pos</code> in the <code>dim</code> dimension of <code>A</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; A = reshape(1:8, 2, 2, 2)
2×2×2 reshape(::UnitRange{Int64}, 2, 2, 2) with eltype Int64:
[:, :, 1] =
 1  3
 2  4

[:, :, 2] =
 5  7
 6  8

julia&gt; sumsliced(A, 2)
14

julia&gt; sumsliced(A, 2, 2)
22</code></pre><p>Heavily inspired by Holy, Tim &quot;Multidimensional algorithms and iteration&quot; <a href="https://julialang.org/blog/2016/02/iteration/#filtering_along_a_specified_dimension_exploiting_multiple_indexes">https://julialang.org/blog/2016/02/iteration/#filtering_along_a_specified_dimension_exploiting_multiple_indexes</a></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/155208e7bd1baa76adff3fee329442e304265b11/src/haplotypecalling.jl#L370-L397">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../haplotype/">« Haplotype Extensions</a><a class="docs-footer-nextpage" href="../private/">Private API Reference »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Sunday 4 June 2023 21:53">Sunday 4 June 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
