<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Pseudoread · HapLink.jl</title><meta name="title" content="Pseudoread · HapLink.jl"/><meta property="og:title" content="Pseudoread · HapLink.jl"/><meta property="twitter:title" content="Pseudoread · HapLink.jl"/><meta name="description" content="Documentation for HapLink.jl."/><meta property="og:description" content="Documentation for HapLink.jl."/><meta property="twitter:description" content="Documentation for HapLink.jl."/><meta property="og:url" content="https://ksumngs.github.io/HapLink.jl/api/psuedoread/"/><meta property="twitter:url" content="https://ksumngs.github.io/HapLink.jl/api/psuedoread/"/><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/api/psuedoread/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../tutorial/1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="../../tutorial/2-examples/">Kicking the tires (Fake sequences)</a></li><li><a class="tocitem" href="../../tutorial/3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="../../tutorial/4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li><a class="tocitem" href="../../cli/variants/">haplink variants</a></li><li><a class="tocitem" href="../../cli/consensus/">haplink consensus</a></li><li><a class="tocitem" href="../../cli/haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="../../cli/sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../variation/">Variation Extensions</a></li><li><a class="tocitem" href="../variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../variationcall/">VariationCall</a></li><li><a class="tocitem" href="../consensus/">Consensus Tools</a></li><li class="is-active"><a class="tocitem" href>Pseudoread</a></li><li><a class="tocitem" href="../haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="../haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Public API Reference</a></li><li class="is-active"><a href>Pseudoread</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Pseudoread</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/api/psuedoread.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="PseudoRead"><a class="docs-heading-anchor" href="#PseudoRead"><code>PseudoRead</code></a><a id="PseudoRead-1"></a><a class="docs-heading-anchor-permalink" href="#PseudoRead" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.Pseudoread" href="#HapLink.Pseudoread"><code>HapLink.Pseudoread</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">Pseudoread(startpos::Integer, endpos::Integer, read::Haplotype)
Pseudoread(query::Union{SAM.Record,BAM.Record}, reference::NucleotideSeq)</code></pre><p>A stand-in struct for a sequencing read (real or imaginary) represented by its position and differences relative to a reference sequence.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L1-L7">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.contradicts-Union{Tuple{T}, Tuple{S}, Tuple{SequenceVariation.Variation{S, T}, SequenceVariation.Haplotype{S, T}}} where {S, T}" href="#HapLink.contradicts-Union{Tuple{T}, Tuple{S}, Tuple{SequenceVariation.Variation{S, T}, SequenceVariation.Haplotype{S, T}}} where {S, T}"><code>HapLink.contradicts</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">contradicts(var::Variation{S,T}, ref::Haplotype{S,T}) where {S,T}</code></pre><p>Returns <code>true</code> if <code>var</code> contains a modification incompatible with any of the <code>Variation</code>s that make up <code>ref</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; using BioSequences, SequenceVariation

julia&gt; ref = Haplotype(dna&quot;GATTACA&quot;, [Variation(dna&quot;GATTACA&quot;, &quot;A2T&quot;), Variation(dna&quot;GATTACA&quot;, &quot;Δ5-6&quot;)]);

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;A2C&quot;), ref)
true

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;T4A&quot;), ref)
false

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;Δ2-2&quot;), ref)
true

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;Δ4-4&quot;), ref)
false

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;2G&quot;), ref)
false

julia&gt; contradicts(Variation(dna&quot;GATTACA&quot;, &quot;4G&quot;), ref)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L257-L287">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.haplotype-Tuple{Pseudoread}" href="#HapLink.haplotype-Tuple{Pseudoread}"><code>HapLink.haplotype</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">haplotype(ps::Pseudoread)</code></pre><p>Gets the differences between <code>ps</code> and its reference sequence as a <code>Haplotype</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L32-L36">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.overlap_inrange-Tuple{Pseudoread, Pseudoread, Integer, Integer}" href="#HapLink.overlap_inrange-Tuple{Pseudoread, Pseudoread, Integer, Integer}"><code>HapLink.overlap_inrange</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">overlap_inrange(
    x::Pseudoread, y::Pseudoread, overlap_min::Integer, overlap_max::Integer
)</code></pre><p>Returns <code>true</code> if the overlap between <code>x</code> and <code>y</code> is within <code>overlap_min</code> and <code>overlap_max</code>, returns <code>false</code> otherwise</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; using BioSequences, SequenceVariation

julia&gt; ps1 = Pseudoread(1, 100, Haplotype(dna&quot;A&quot;, Variation{LongDNA{4}, DNA}[]));

julia&gt; ps2 = Pseudoread(50, 150, Haplotype(dna&quot;A&quot;, Variation{LongDNA{4}, DNA}[]));

julia&gt; overlap_inrange(ps1, ps2, 1, 500)
true

julia&gt; overlap_inrange(ps1, ps2, 75, 500)
false

julia&gt; overlap_inrange(ps1, ps2, 1, 25)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L310-L335">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.overlapping_variations-Tuple{Pseudoread, SequenceVariation.Haplotype}" href="#HapLink.overlapping_variations-Tuple{Pseudoread, SequenceVariation.Haplotype}"><code>HapLink.overlapping_variations</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">overlapping_variations(ps::Pseudoread, v::Haplotype)</code></pre><p>Find all <code>Variation</code>s within <code>v</code> that are contained within the range defined by <code>ps</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L296-L300">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.pseudoreads-Tuple{Union{AbstractString, FilePathsBase.AbstractPath}, BioSequences.BioSequence{&lt;:BioSequences.NucleicAcidAlphabet}}" href="#HapLink.pseudoreads-Tuple{Union{AbstractString, FilePathsBase.AbstractPath}, BioSequences.BioSequence{&lt;:BioSequences.NucleicAcidAlphabet}}"><code>HapLink.pseudoreads</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">pseudoreads(sam::Union{AbstractString,AbstractPath}, consensus::NucleotideSeq)</code></pre><p>Create a set of <a href="#HapLink.Pseudoread"><code>Pseudoread</code></a>s from the alignments present in <code>sam</code> when aligned against <code>consensus</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L39-L44">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.simulate-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{Pseudoread}, BioSequences.BioSequence, AbstractArray{SequenceVariation.Variation{S, T}}}} where {S, T}" href="#HapLink.simulate-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{Pseudoread}, BioSequences.BioSequence, AbstractArray{SequenceVariation.Variation{S, T}}}} where {S, T}"><code>HapLink.simulate</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">simulate(
    pool::AbstractArray{Pseudoread},
    refseq::BioSequence,
    subcon::AbstractArray{Variation{S,T}};
    reverse_order::Union{Bool,Nothing}=nothing,
    overlap_min::Int64=0,
    overlap_max::Int64=500,
) where {S,T} -&gt; Union{Haplotype{S,T},Nothing}</code></pre><p>Creates a new set of <a href="#HapLink.Pseudoread"><code>Pseudoread</code></a>s based on the <code>pool</code> of real reads spliced spliced together using maximum likelihood simulation.</p><p><strong>Arguments</strong></p><ul><li><code>pool::AbstractArray{Pseudoread}</code>: A set of (presumably) real sequencing reads that will serve as pieces of the new simulated reads.</li><li><code>refseq::BioSequence</code>: The reference against which all of the sequences in <code>pool</code> and <code>subcon</code> were aligned</li><li><code>subcon::AbstractArray{Variation{S,T}}</code>: The subconsensus <code>Variation</code>s which are known to be both present in <code>pool</code> and real (i.e., not due to sequencing errors)</li></ul><p><strong>Options</strong></p><ul><li><code>reverse_order::Union{Bool,Nothing}=nothing</code>: Whether the splicing should start at the beginning or end of the reference sequence. If left blank, will randomly decide.</li><li><code>overlap_min::Int64=0</code>: The amount that reads from <code>pool</code> are required to overlap in order to be spliced together. Can be negative, indicating that reads must be spaced apart by this amount, instead. -<code>overlap_max::Int64=500</code>: The amount that reads from pool can overlap before being discarded. Like <code>overlap_min</code>, can be negative.</li></ul><p><strong>Returns</strong></p><p>A <code>Haplotype{S,T}</code> representing the simulated read, or <code>nothing</code> if the simulation encountered a dead end.</p><p><strong>Extended help</strong></p><p>The simulation procedure works by picking a read from <code>pool</code> that contains the first (if <code>reverse_order</code> is false, the last otherwise) variant from <code>subcon</code>. The procedure examines every position in <code>subcon</code> in ascending (or descending, if <code>reverse_order == true</code>) order until the chosen read no longer carries that position. The simulation will then, at random, pick a read from <code>pool</code> that contains the next position from <code>subcon</code> and that also <em>has matching variations at every position in <code>subcon</code> as the previous read</em>, as well as meets the overlap requirements. These two reads may differ in sites other than those in <code>subcon</code>: it is assumed that appropriate variant calling has already been performed and that any variation in those sites is due to sequencing error. The procedure repeats for each new read from <code>pool</code> until the simulation has simulated every position of <code>subcon</code>, or until it reaches a dead-end and cannot find a read that matches every requirement.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L59-L109">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="HapLink.variations_match-Tuple{Pseudoread, SequenceVariation.Haplotype}" href="#HapLink.variations_match-Tuple{Pseudoread, SequenceVariation.Haplotype}"><code>HapLink.variations_match</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">variations_match(reference::Pseudoread, query::Union{Pseudoread,Haplotype})</code></pre><p>Returns <code>true</code> if <code>query</code> could be a read from <code>reference</code>.</p><p>Additional <code>Variation</code>s found within <code>query</code> that are not present in <code>reference</code> <strong>do not</strong> invalid a positive match result so long as none of those <code>Variation</code>s <a href="#HapLink.contradicts-Union{Tuple{T}, Tuple{S}, Tuple{SequenceVariation.Variation{S, T}, SequenceVariation.Haplotype{S, T}}} where {S, T}"><code>contradicts</code></a> <code>reference</code>. These <code>Variation</code>s can either</p><ol><li>Extend beyond the length of <code>reference</code>, or</li><li>Exist as &quot;sequencing errors&quot; within the interval of <code>reference</code></li></ol><p>On the other hand, <strong>every</strong> <code>Variation</code> within the overlapping interval of <code>reference</code> and <code>query</code> that is present in <code>reference</code> <strong>must</strong> also be found in <code>query</code>. In other words, <code>query</code> must have all of <code>reference</code> (within the overlap), but <code>reference</code> does not need all of <code>query</code>.</p><p>This arrangment allows for the creation of bodies of matching <a href="#HapLink.Pseudoread"><code>Pseudoread</code></a>s, as done via <a href="#HapLink.simulate-Union{Tuple{T}, Tuple{S}, Tuple{AbstractArray{Pseudoread}, BioSequences.BioSequence, AbstractArray{SequenceVariation.Variation{S, T}}}} where {S, T}"><code>simulate</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/676eb6a7eb7c5140f2eb325a6b35bc87108eae08/src/pseudoread.jl#L216-L235">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../consensus/">« Consensus Tools</a><a class="docs-footer-nextpage" href="../haplotype/">Haplotype Extensions »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Friday 22 December 2023 19:32">Friday 22 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
