<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>haplink variants · HapLink.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://ksumngs.github.io/HapLink.jl/cli/variants/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img class="docs-light-only" src="../../assets/logo.png" alt="HapLink.jl logo"/><img class="docs-dark-only" src="../../assets/logo-dark.png" alt="HapLink.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">HapLink.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Tutorials</span><ul><li><a class="tocitem" href="../../tutorial/1-install/">In the beginning (Installation)</a></li><li><a class="tocitem" href="../../tutorial/2-examples/">Kicking the tires (Fake sequences)</a></li><li><a class="tocitem" href="../../tutorial/3-other/">Plays well with others (External tools)</a></li><li><a class="tocitem" href="../../tutorial/4-repl/">For advanced beginners (REPL mode)</a></li></ul></li><li><span class="tocitem">CLI Reference</span><ul><li class="is-active"><a class="tocitem" href>haplink variants</a></li><li><a class="tocitem" href="../consensus/">haplink consensus</a></li><li><a class="tocitem" href="../haplotypes/">haplink haplotypes</a></li><li><a class="tocitem" href="../sequences/">haplink sequences</a></li></ul></li><li><span class="tocitem">Public API Reference</span><ul><li><a class="tocitem" href="../../api/variation/">Variation Extensions</a></li><li><a class="tocitem" href="../../api/variationinfo/">VariationInfo</a></li><li><a class="tocitem" href="../../api/variationpileup/">VariationPileup</a></li><li><a class="tocitem" href="../../api/variationcall/">VariationCall</a></li><li><a class="tocitem" href="../../api/consensus/">Consensus Tools</a></li><li><a class="tocitem" href="../../api/psuedoread/">Pseudoread</a></li><li><a class="tocitem" href="../../api/haplotype/">Haplotype Extensions</a></li><li><a class="tocitem" href="../../api/haplotypecall/">HaplotypeCall</a></li></ul></li><li><a class="tocitem" href="../../api/private/">Private API Reference</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">CLI Reference</a></li><li class="is-active"><a href>haplink variants</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>haplink variants</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/ksumngs/HapLink.jl/blob/master/docs/src/cli/variants.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="variants-cli"><a class="docs-heading-anchor" href="#variants-cli"><code>haplink variants</code></a><a id="variants-cli-1"></a><a class="docs-heading-anchor-permalink" href="#variants-cli" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-binding" id="HapLink.variants" href="#HapLink.variants"><code>HapLink.variants</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">haplink variants [options] reference bam</code></pre><p>Call variants</p><p><strong>Introduction</strong></p><p>Decides which variations found within an alignment are real, and which are due to random chance. HapLink uses Fisher&#39;s Exact Test to determine the statistical significance of sequence variations, and optionally allows for other thresholds to reduce random noise in the variant calling. Outputs a Variant Call Format (VCF) file compliant with VCF v4.</p><p><strong>Arguments</strong></p><ul><li><code>reference</code>: path to the reference genome to call variants against in fasta format. Must   not be gzipped, but does not need to be indexed (have a sidecar fai file). HapLink only   supports single-segment reference genomes: if <code>reference</code> includes more than one   sequence, all but the first will be ignored.</li><li><code>bam</code>: alignment file to call variants from. Can be in SAM or BAM format, and does not   need to be sorted or indexed, but variant calling speed will increase significantly if   using a sorted and indexed (has a sidebar bai file) BAM file.</li></ul><p><strong>Options</strong></p><ul><li><code>--outfile=&lt;path&gt;</code>: The file to write variant calls to. If left blank, variant calls are   written to standard output.</li><li><code>--significance=&lt;float&gt;</code>: The alpha value for statistical significance of variant calls.</li><li><code>--depth=&lt;int&gt;</code>: Minimum number of times the variation must be observed within the   alignment to be called a variant</li><li><code>--quality=&lt;float&gt;</code>: The minimum average basecall quality score for a variation to be   called a variant</li><li><code>--frequency=&lt;float&gt;</code>: The minimum proportion of reads that the variation must be observed   within compared to all reads covering its position for that variation to be called a   variant</li><li><code>--position=&lt;float&gt;</code>: The distance (as a percentage) from the edge of reads that a   variation must be observed at to be called a variant</li><li><code>--strandedness=&lt;float&gt;</code>: The maximum proportion of times that a variation can be observed   on one strand versus the other to be called a variant. This metric is totally useless on   single-stranded sequencing protocols like Oxford Nanopore, but can be useful for   combining data between stranded protocols like most Illumina and Pacific Bio.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/ksumngs/HapLink.jl/blob/85c79a0636a5b0be2ccdbc7d0c4580a1eaf6581d/src/HapLink.jl#L104-L144">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../tutorial/4-repl/">« For advanced beginners (REPL mode)</a><a class="docs-footer-nextpage" href="../consensus/">haplink consensus »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Wednesday 28 June 2023 22:08">Wednesday 28 June 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
